lib/liquid/static_registers.rb

```diff
@@ -0,0 +1,36 @@
+module Liquid
+  class StaticRegisters
+    def initialize(registers = {})
+      @static_registers = registers.is_a?(StaticRegisters) ? registers.static : registers.freeze
+      @registers = {}
+    end
+
+    def []=(key, value)
+      @registers[key] = value
+    end
+
+    def [](key)
+      if @registers.key?(key)
+        @registers[key]
+      else
+        @static_registers[key]
+      end
+    end
+
+    def delete(key)
+      @registers.delete(key)
+    end
+
+    def fetch(key, default = nil)
+      self[key] || default
```

Comment:
I think the proper implementation is `key?(key) ? self[key] : default` otherwise we could return `default` when `self[key]` has been assigned `nil` or `false`.